xdsl/backend/riscv/register_allocation.py

```diff
@@ -81,6 +100,15 @@ def _free(self, reg: SSAValue) -> None:
             self.available_registers.push(reg.type)
 
     def allocate_func(self, func: riscv_func.FuncOp) -> None:
+        if self.exclude_preallocated:
+            preallocated = _gather_allocated(func)
+
+            for pa_reg in preallocated:
```

Comment:
We probably want to add preallocated registers to `available_registers.reserved_registers` so that we don't accidentally free them

Reply:
That's a good suggestion, but I'd argue that we need a different set for tracking this.
I think reserved registers have had a different meaning so far.
What do you think?

Reply:
maybe it's a docs/naming thing? In the RegisterQueue, functionally, it's the set of registers that cannot be pushed into the queue, which seems to fit the requirements here